### PR TITLE
provide input_model for nnet3 training scripts and libs

### PR DESCRIPTION
**Pull request by naxingyu**
As per discussion in [forum](https://groups.google.com/forum/#!msg/kaldi-help/yEW6OXZ2mjY/Cbihb4LlBgAJ).
The update for frame_obj training lib is included in the aishell2 PR #2398 as a merged [commit](https://github.com/kaldi-asr/kaldi/pull/2398/commits/cc62e8a48ff50b63adeca5283389167fbbb0df11).

### CODE REVIEW

**Comment by naxingyu** [comment]
@danpovey

**Comment by danpovey** [comment]
Thanks.  Have you tested that a regular nnet3 training run (i.e. without this option) works?  I want to make sure the old setup does not break.

**Comment by naxingyu** [comment]
Yes, @underdogliu used this script while validating the aishell2 results.

File: egs/wsj/s5/steps/libs/nnet3/train/common.py [hunk 1]
```diff
@@ -535,7 +535,10 @@ def smooth_presoftmax_prior_scale_vector(pdf_counts,
     return scaled_counts
 
 
-def prepare_initial_network(dir, run_opts, srand=-3):
+def prepare_initial_network(dir, run_opts, srand=-3, input_model=None):
+    if input_model is not None:
+        shutil.copy2(input_model, "{0}/0.raw".format(dir))
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/train/common.py]: it seems strange to me to use copy2 (which copies timestamps etc.)  I think this should just be 'copy'.

Comment (naxingyu) [reply egs/wsj/s5/steps/libs/nnet3/train/common.py]: I referred to [this](https://github.com/naxingyu/kaldi/blob/master/egs/wsj/s5/steps/libs/nnet3/train/common.py#L323) line of code. Should I change both?

**Comment by danpovey** [comment]
yes please.

**Comment by naxingyu** [comment]
done

Best,
Xingyu Na

**Comment by naxingyu** [comment]
I see. Thanks!
Best,
Xingyu Na

**Comment by xiaosdawn** [comment]
Hi, everyone,
I have studied the finetune_tdnn.sh in aishell2 to finetune my Chain model. But there is some error bothers me. Could anyone do the same thing ? or help me to resolve this problem ? 
Any advises or ideas will be appreciate! Thanks.

The error information is in the train.0.*.log,  detail is :

# nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --write-cache=exp/chain/tdnn_sp_v0_finetune_chain/cache.1 --xent-regularize=0.1 --print-interval=10 --momentum=0.0 --max-param-change=1.41421356237 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.5 --srand=0 "nnet3-am-copy --raw=true --learning-rate=0.002 --scale=1.0 exp/chain/tdnn_sp_v0_finetune_chain/0.mdl - |" exp/chain/tdnn_sp_v0_finetune_chain/den.fst "ark,bg:nnet3-chain-copy-egs --frame-shift=1                     ark:exp/chain/tdnn_sp_v0_finetune_chain/egs/cegs.1.ark ark:- | nnet3-chain-shuffle-egs --buffer-size=5000 --srand=0 ark:- ark:- | nnet3-chain-merge-egs --minibatch-size=64 ark:- ark:- |" exp/chain/tdnn_sp_v0_finetune_chain/1.1.raw 
# Started at Wed Aug 15 17:02:47 CST 2018
#
nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm coefficient=0.1 --write-cache=exp/chain/tdnn_sp_v0_finetune_chain/cache.1 --xent-regularize=0.1 --print-interval=10 --momentum=0.0 --max-param-change=1.41421356237 --backstitch-training-scale=0.0 --backstitch-training-interval=1 --l2-regularize-factor=0.5 --srand=0 'nnet3-am-copy --raw=true --learning-rate=0.002 --scale=1.0 exp/chain/tdnn_sp_v0_finetune_chain/0.mdl - |' exp/chain/tdnn_sp_v0_finetune_chain/den.fst 'ark,bg:nnet3-chain-copy-egs --frame-shift=1                         ark:exp/chain/tdnn_sp_v0_finetune_chain/egs/cegs.1.ark ark:- | nnet3-chain-shuffle-egs --buffer-size=5000 --srand=0 ark:- ark:- | nnet3-chain-merge-egs --minibatch-size=64 ark:- ark:- |' exp/chain/tdnn_sp_v0_finetune_chain/1.1.raw

LOG (nnet3-chain-train[5.4]:SelectGpuId():cu-device.cc:191) CUDA setup operating under Compute Exclusive Mode.
LOG (nnet3-chain-train[5.4]:FinalizeActiveGpu():cu-device.cc:247) The active GPU is [1]: GeForce GTX 1080 Ti    free:10947M, used:225M, total:11172M, free/total:0.979861 version 6.1
nnet3-am-copy --raw=true --learning-rate=0.002 --scale=1.0 exp/chain/tdnn_sp_v0_finetune
_chain/0.mdl -
LOG (nnet3-am-copy[5.4]:main():nnet3-am-copy.cc:151) Copied neural net from exp/chain/td
nn_sp_v0_finetune_chain/0.mdl to raw format as -
nnet3-chain-shuffle-egs --buffer-size=5000 --srand=0 ark:- ark:-
nnet3-chain-copy-egs --frame-shift=1 ark:exp/chain/tdnn_sp_v0_finetune_chain/egs/cegs.1.ark ark:-
nnet3-chain-merge-egs --minibatch-size=64 ark:- ark:-
LOG (nnet3-chain-copy-egs[5.4]:main():nnet3-chain-copy-egs.cc:395) Read 363 neural-network training examples, wrote 363
ASSERTION_FAILED (nnet3-chain-train[5.4]:NumeratorComputation():chain-numerator.cc:36) : 'supervision.num_sequences * supervision.frames_per_sequence == nnet_output.NumR
ows() && supervision.label_dim == nnet_output.NumCols()'

[ Stack-Trace: ]
nnet3-chain-train() [0x126989c]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::KaldiAssertFailure_(char const*, char const*, int, char const*)
kaldi::chain::NumeratorComputation::NumeratorComputation(kaldi::chain::Supervision const&, kaldi::CuMatrixBase<float> const&)
kaldi::chain::ComputeChainObjfAndDeriv(kaldi::chain::ChainTrainingOptions const&, kaldi::chain::DenominatorGraph const&, kaldi::chain::Supervision const&, kaldi::CuMatri
xBase<float> const&, float*, float*, float*, kaldi::CuMatrixBase<float>*, kaldi::CuMatrix<float>*)
kaldi::nnet3::NnetChainTrainer::ProcessOutputs(bool, kaldi::nnet3::NnetChainExample const&, kaldi::nnet3::NnetComputer*)
kaldi::nnet3::NnetChainTrainer::TrainInternal(kaldi::nnet3::NnetChainExample const&, kaldi::nnet3::NnetComputation const&)
kaldi::nnet3::NnetChainTrainer::Train(kaldi::nnet3::NnetChainExample const&)
main
__libc_start_main
_start

ERROR (nnet3-chain-merge-egs[5.4]:Write():kaldi-matrix.cc:1403) Failed to write matrix to stream
<in-3700h16k-tone_tri4b_SATLDAMLLT_news-speech-for-jindiaoyi.4gram-tone-addedPronSil/tdnn_sp_v0_finetune_meeting_chain/log/train.0.1.log" 107L, 10448C 1,1

**Comment by danpovey** [comment]
Looks to me like a mismatch in nnet output dim, which would be caused by
some kind of mismatch in tree.  That script (I'm not 100% sure but I'm
guessing) probably expects the tree to be the same.

**Comment by xiaosdawn** [comment]
Danpovey,
Thank you very much. And your guessing is right. There is a mismatch between nnet output dim and tree, due to my misunderstanding with the detail for finetuning. 
Now the problem is resoveled, when I employ the tree that was used to train my Chain model.

**Comment by naxingyu** [comment]
There is not instruction on how to use that finetuning script. We'll update
later. Thanks!
Best,
Xingyu Na
